File: Data Structures & Algorithms/maximum-product-subarray/submission-1.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n = len(nums)
        smallest, largest, res = nums[0], nums[0], nums[0]

        for i in range(1, n):
            smallest *= nums[i]
            largest *= nums[i]
            tmp = smallest
            smallest = min(smallest, largest, nums[i])
            largest = max(tmp, largest, nums[i])
            res = max(res, largest)
        return res

File: Data Structures & Algorithms/max-water-container/submission-1.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int ans = 0;
        int n = heights.size();
        int l = 0;
        int r = n-1;

        while (l < r) {
            int area = (min(heights[l], heights[r]) * (r-l));
            ans = max(ans, area);
            if (heights[l] > heights[r]) {
                r--;
            }
            else {
                l++;
            }
        }
        return ans;
    }
};
